pyvisfile/xdmf/__init__.py

```diff
@@ -0,0 +1,1169 @@
+        """
+
+        super().__init__(parent, "Geometry", {
+            "Name": name,
+            "GeometryType": gtype.name,
+            })
+
+# }}}
+
+
+# {{{ time
+
+class TimeType(enum.Enum):
+    """Temporal information of the current grid."""
+    Single = enum.auto()
+    HyperSlab = enum.auto()
+    List = enum.auto()
+    Range = enum.auto()
+
+
+class Time(XdmfElement):
+    """
+    .. automethod:: __init__
+    """
+
+    def __init__(
+            self, *,
+            value: str,
+            parent: Optional[Element] = None,
+            ):
+        """
+        :param parent: if provided, *self* is appended to the element.
+        """
+
+        super().__init__(parent, "Time", {
+            "Value": value,
+            })
+
+# }}}
+
+
+# {{{ domain
+
+class Domain(XdmfElement):
+    """
+    .. automethod:: __init__
+    """
+
+    def __init__(
+            self, *,
+            name: Optional[str] = None,
+            parent: Optional[Element] = None,
+            ):
+        """
+        :param parent: if provided, *self* is appended to the element.
+        """
+
+        super().__init__(parent, "Domain", {
+            "Name": name,
+            })
+
+# }}}
+
+
+# {{{ information
+
+class Information(XdmfElement):
+    """
+    .. automethod:: __init__
+    """
+
+    def __init__(
+            self, *,
+            name: str,
+            value: str,
+            parent: Optional[Element] = None,
+            ):
+        """
+        :param parent: if provided, *self* is appended to the element.
+        """
+
+        super().__init__(parent, "Information", {
+            "Name": name,
+            "Value": value,
+            })
+
+# }}}
+
+
+# {{{ include
+
+class XInclude(XdmfElement):
+    """
+    .. automethod:: __init__
+    """
+
+    def __init__(
+            self, *,
+            href: Optional[str],
+            xpointer: Optional[str] = None,
+            parent: Optional[Element] = None,
+            ):
+        """
+        :param parent: if provided, *self* is appended to the element.
+        :param xpointer: path inside the file represented by *href*.
+        """
+        if xpointer is not None:
+            xpointer = f"xpointer({xpointer})"
+
+        super().__init__(parent, "xi:include", {
+            "href": href,
+            "xpointer": xpointer,
+            })
+
+# }}}
+
+# }}}
+
+
+# {{{ data arrays
+
+def _ndarray_to_string(ary):
+    if not isinstance(ary, np.ndarray):
+        raise TypeError(f"expected an 'ndarray', got '{type(ary).__name__}'")
+
+    ntype = _numpy_to_xdmf_number_type(ary.dtype)
+    if ntype == DataItemNumberType.Int or ntype == DataItemNumberType.UInt:
+        fmt = "%d"
+    elif ntype == DataItemNumberType.Float:
+        fmt = "%.16e" if ary.dtype.itemsize == 8 else "%.8e"
+    else:
+        raise ValueError(f"unsupported dtype: '{ary.dtype}'")
+
+    import io
+    bio = io.BytesIO()
+    np.savetxt(bio, ary, fmt=fmt)
```

Comment:
I really like Python's `repr` for this (at least for DP), because it guarantees that DP numbers round-trip exactly. Any chance that can be used here?

Reply:
None that I could find, unfortunately. 

[savetxt](https://numpy.org/doc/stable/reference/generated/numpy.savetxt.html?highlight=savetxt#numpy.savetxt) just uses a printf-style formatting string. Other things like [array2string](https://numpy.org/doc/stable/reference/generated/numpy.array2string.html) seem to be able to do something like that with the `formatter` or `floatmode` flags, but they just give the repr of the `ndarray` with brackets, which is not usable here.

If it helps, this is what [meshio](https://github.com/nschloe/meshio/blob/37673c8fb938ad73d92fb3171dee3eb193b5e7ac/meshio/xdmf/common.py#L20) is doing too.